### PR TITLE
dev-util/hip: Include test suite

### PR DESCRIPTION
**Pull request by heroxbd**
The dev-util/hip upstream has separated its test cases into a full git
repository with an independent build system.  To run the test cases,
we have to call cmake twice in every src_* phase. I come up with a
wrapper to hack CMAKE_USE_DIR and BUILD_DIR values.

I am not sure if this is the recommended way to use `cmake.eclass`.
Therefore this PR.

### CODE REVIEW

**Comment by littlewu2508** [comment]
Thanks for finding a simple solution for building and running tests! (rather than tremendous patch which re-bundle the test repo in my own experimental implementation at https://github.com/littlewu2508/gentoo/commit/c368f52c3770d468a5e7698db044b677b0144feb)

Can you show a example test log? As I remember https://bugs.gentoo.org/916069 can cause test failure.

**Comment by heroxbd** [comment]
> Can you show an example test log? As I remember https://bugs.gentoo.org/916069 can cause test failure.

This should be run after https://github.com/gentoo/gentoo/pull/33900.  Sorry I have not saved my test log.

**Comment by heroxbd** [comment]
I find ebuilds calling cmake_src_compile twice:

```
games-action/minetest
games-engines/openmw
games-simulation/corsix-th
media-libs/shaderc
media-sound/drumstick
media-sound/hydrogen
media-sound/zynaddsubfx
net-irc/kvirc
net-libs/davix
net-libs/libnetconf2
net-libs/libssh
net-libs/libyang
net-misc/sysrepo
sci-astronomy/celestia
sci-chemistry/molequeue
sci-electronics/kicad
sci-geosciences/mapserver
sci-libs/ViSP
sci-libs/fcl
sci-libs/gdal
sci-libs/geos
sci-libs/liborigin
sci-libs/nlopt
sci-libs/scalapack
sys-libs/pam_wrapper
app-arch/createrepo_c
app-doc/doxygen
app-office/ledger
app-text/sdcv
dev-games/mygui
dev-games/openscenegraph
dev-games/openscenegraph-openmw
dev-games/physfs
dev-lang/yap
dev-libs/grantlee
dev-libs/libical
dev-util/cmocka
```
My approach is likely to be valid.

**Comment by littlewu2508** [comment]
> I find ebuilds calling cmake_src_compile twice:
> 
> My approach is likely to be valid.

I think your approach is good, though I don't think most of the above examples do the same. Your trick is reset cmake related variables and recall every `cmake_src_` functions; while the examples you listed have different senarios, including:

1. There are plenty targets to make, so call multiple `cmake_src_compile <extra_target_name>`, thus it just have one `${CMKAE_USE_DIR}`, one `${BUILD_DIR}`, one `cmake_src_configure`, just multiple targets. E.g. `app-doc/doxygen`
2. There are different python implementations to build, thus it has just one `${CMKAE_USE_DIR}`, one `cmake_src_prepare`, `python_foreach_impl cmake_src_configure` which means multiple `${BUILD_DIR}`. E.g. `sys-libs/pam_wrapper` `sci-geosciences/mapserver`
3. Provide a static build, so it has two set of `${BUILD_DIR}` and configurations (two `cmake_src_configure`), but one `${CMKAE_USE_DIR}`. E.g. `dev-games/mygui`

So after filtering, there are 2 packages that have similar approaches with yours:

`sci-libs/dlib` and `sci-astronomy/celestia`  but neither of them call `cmake_src_prepare` twice.

`sci-libs/dlib` put `cmake_src_configure` and `cmake_src_compile` inside `src_test` to compile the test project, while I prefer your way.

`sci-astronomy/celestia` seems to have a sub project to compile so it calls `cmake_src_configure` and `cmake_compile` twice (and I believe it should call `cmake_src_prepare` twice as well)

So in conclusion I think your way is valid, but not because it's existing practice in ::gentoo, but because it sounds better than the existing practice in `sci-libs/dlib` and `sci-astronomy/celestia`

**Comment by littlewu2508** [comment]
@heroxbd 

I got

```
FAILED: catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o                                                                                                                                                       
/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/clr-rocm-5.7.1_build/hipamd/bin/hipcc -DKERNELS_PATH=\"/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/kernels/\" -I/run/user/18014/portage/dev-util/hip-5.7.
1-r1/work/hip-tests-rocm-5.7.0/catch/external/Catch2 -I/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/./include -I/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/./kernels -I/ru
n/user/18014/portage/dev-util/hip-5.7.1-r1/work/clr-rocm-5.7.1_build/hipamd/include -I/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/external/picojson  -O2 -pipe --rocm-path=/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/clr-rocm-5.7.1_build/hipamd --std=c++17 --offload-arch=gfx90a   -Xclang -fallow-half-arguments-and-returns -MD -MT catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o -MF catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o.d -o catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o -c /run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/unit/deviceLi
b/hipMathFunctions.cc                                                                                                                                                                                                                         
error: unknown argument: '-fallow-half-arguments-and-returns'
```

when compiling the tests.

Perhaps you need https://github.com/littlewu2508/gentoo/commit/c368f52c3770d468a5e7698db044b677b0144feb#diff-8b8998ee2c8e3f167539d22040daa4679a7190ca8285e9acf2824a1eadee6941 ?

File: eclass/rocm.eclass [hunk 1]
```diff
@@ -176,9 +176,12 @@ _rocm_set_globals() {
 	local optflags=${allflags[@]/%/(-)?}
 	ROCM_USEDEP=${optflags// /,}
 }
-_rocm_set_globals
-unset -f _rocm_set_globals
 
+# dev-util/hip uses checkgpu but does not need the USE flags.
```
Comment (littlewu2508) [inline eclass/rocm.eclass:180]: Actually there's `dev-libs/rocm-opencl-runtime` which also need just `check-amdgpu` but not other global variables. Its approach is simply copy the `check-amdgpu` to the ebuild. Here you provide an alternative way, but restrict to `dev-util/hip`. I think it would be better to let each package decide if they want to inherit `rocm.eclass` without global variables.

File: eclass/rocm.eclass [hunk 2]
```diff
@@ -176,9 +176,12 @@ _rocm_set_globals() {
...
+if [[ ${CATEGORY}/${PN} != dev-util/hip ]]; then
```
Comment (littlewu2508) [inline eclass/rocm.eclass:181]: Something like, and insert a `ROCM_ECLASS_NO_GLOBALS=true` before `inherit rocm.eclass` in `dev-util/hip-5.7.1-r1.ebuild`

```suggestion
if [[ -n ${ROCM_ECLASS_NO_GLOBALS} ]]; then
```

**Comment by littlewu2508** [comment]
> @heroxbd
> 
> I got
> 
> ```
> FAILED: catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o                                                                                                                                                       
> /run/user/18014/portage/dev-util/hip-5.7.1-r1/work/clr-rocm-5.7.1_build/hipamd/bin/hipcc -DKERNELS_PATH=\"/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/kernels/\" -I/run/user/18014/portage/dev-util/hip-5.7.
> 1-r1/work/hip-tests-rocm-5.7.0/catch/external/Catch2 -I/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/./include -I/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/./kernels -I/ru
> n/user/18014/portage/dev-util/hip-5.7.1-r1/work/clr-rocm-5.7.1_build/hipamd/include -I/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/external/picojson  -O2 -pipe --rocm-path=/run/user/18014/portage/dev-util/hip-5.7.1-r1/work/clr-rocm-5.7.1_build/hipamd --std=c++17 --offload-arch=gfx90a   -Xclang -fallow-half-arguments-and-returns -MD -MT catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o -MF catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o.d -o catch_tests/unit/deviceLib/CMakeFiles/UnitDeviceTests.dir/hipMathFunctions.cc.o -c /run/user/18014/portage/dev-util/hip-5.7.1-r1/work/hip-tests-rocm-5.7.0/catch/unit/deviceLi
> b/hipMathFunctions.cc                                                                                                                                                                                                                         
> error: unknown argument: '-fallow-half-arguments-and-returns'
> ```
> 
> when compiling the tests.
> 
> Perhaps you need [littlewu2508@c368f52#diff-8b8998ee2c8e3f167539d22040daa4679a7190ca8285e9acf2824a1eadee6941](https://github.com/littlewu2508/gentoo/commit/c368f52c3770d468a5e7698db044b677b0144feb#diff-8b8998ee2c8e3f167539d22040daa4679a7190ca8285e9acf2824a1eadee6941) ?

Backporting https://reviews.llvm.org/rG7f12dcac79f3d2653ae77cd1be0768e64ff7a518 to clang-17 would be a better choice.

Ref: https://github.com/ROCm-Developer-Tools/HIP/issues/3178

**Comment by heroxbd** [comment]
littlewu2508 ***@***.***> writes:

> I think your approach is good, though I don't think most of the above examples do the same. Your trick is reset cmake related variables
> and recall every cmake_src_ functions; while the examples you listed have different senarios, including:
>
> 1 There are plenty targets to make, so call multiple cmake_src_compile <extra_target_name>, thus it just have one ${CMKAE_USE_DIR},
>  one ${BUILD_DIR}, one cmake_src_configure, just multiple targets. E.g. app-doc/doxygen
> 2 There are different python implementations to build, thus it has just one ${CMKAE_USE_DIR}, one cmake_src_prepare,
>  python_foreach_impl cmake_src_configure which means multiple ${BUILD_DIR}. E.g. sys-libs/pam_wrapper
>  sci-geosciences/mapserver
> 3 Provide a static build, so it has two set of ${BUILD_DIR} and configurations (two cmake_src_configure), but one ${CMKAE_USE_DIR}. E.g.
>  dev-games/mygui
>
> So after filtering, there are 2 packages that have similar approaches with yours:
>
> sci-libs/dlib and sci-astronomy/celestia but neither of them call cmake_src_prepare twice.
>
> sci-libs/dlib put cmake_src_configure and cmake_src_compile inside src_test to compile the test project, while I prefer your way.
>
> sci-astronomy/celestia seems to have a sub project to compile so it calls cmake_src_configure and cmake_compile twice (and I believe it
> should call cmake_src_prepare twice as well)
>
> So in conclusion I think your way is valid, but not because it's existing practice in ::gentoo, but because it sounds better than the existing
> practice in sci-libs/dlib and sci-astronomy/celestia

Yiyang, thank you for the comprehensive survey.  It's very helpful!

File: eclass/rocm.eclass (hunk 1 above)
Comment (heroxbd) [reply eclass/rocm.eclass:180]: Good point.  Thanks!
